invenio_vocabularies/contrib/names/datastreams.py

```diff
@@ -67,42 +68,55 @@ def _process_lambda_file(self, fileobj):
         if self.since:
             time_shift = self.since
         last_sync = arrow.now() - timedelta(**time_shift)
+        try:
+            file_content = fileobj.read().decode("utf-8")
 
-        file_content = fileobj.read().decode("utf-8")
-
-        csv_reader = csv.DictReader(file_content.splitlines())
+            csv_reader = csv.DictReader(file_content.splitlines())
 
-        for row in csv_reader:  # Skip the header line
-            orcid = row["orcid"]
+            for row in csv_reader:  # Skip the header line
+                orcid = row["orcid"]
 
-            # Lambda file is ordered by last modified date
-            last_modified_str = row["last_modified"]
-            try:
-                last_modified_date = arrow.get(last_modified_str, date_format)
-            except arrow.parser.ParserError:
-                last_modified_date = arrow.get(last_modified_str, date_format_no_millis)
+                # Lambda file is ordered by last modified date
+                last_modified_str = row["last_modified"]
+                try:
+                    last_modified_date = arrow.get(last_modified_str, date_format)
+                except arrow.parser.ParserError:
+                    last_modified_date = arrow.get(
+                        last_modified_str, date_format_no_millis
+                    )
 
-            if last_modified_date < last_sync:
-                break
-            yield orcid
+                if last_modified_date < last_sync:
+                    break
+                yield orcid
+        finally:
+            fileobj.close()
 
     def _iter(self, orcids):
         """Iterates over the ORCiD records yielding each one."""
         with ThreadPoolExecutor(
             max_workers=current_app.config["VOCABULARIES_ORCID_SYNC_MAX_WORKERS"]
         ) as executor:
-            futures = [
-                executor.submit(
+            # futures is a dictionary where the key is the ORCID value and the item is the Future object
+            futures = {
+                orcid: executor.submit(
                     self._fetch_orcid_data,
                     orcid,
                     current_app.config["VOCABULARIES_ORCID_SUMMARIES_BUCKET"],
                 )
                 for orcid in orcids
-            ]
-            for future in as_completed(futures):
-                result = future.result()
-                if result is not None:
-                    yield result
+            }
+
+            for orcid, future in list(futures.items()):
```

Comment:
is casting to list needed?

Reply:
Could also be just iterating of `list(futures.keys())`, since the reason is that we're modifying the `futures` dictionary as we iterate over the keys.